src/compactor_state.rs

```diff
@@ -106,11 +108,142 @@ impl CompactorState {
                 return Err(SlateDBError::InvalidCompaction);
             }
         }
+
+        let existing_sst_ids = self
+            .db_state
```

Comment:
Can the logic be simpler?
Here are the properties that we have
* In `CoreDbState`, `l0`  and `compacted` are both ordered.  `l0` + `compacted` is logically a single ordered list.
* In `Compaction`, `sources` is ordered.

With that, if we take the first item in `sources` and find its position in the logical `l0` + `compacted` list, rest of `sources` must have the same sequence of elements as the `l0` + `compacted`.

For destination, it must either be the same as last element in `sources` (and a sr) or is the next new sr number.

Reply:
@brayanjuls , shall we discuss above before further review?

Reply:
@vigneshc thanks for the observations. I tried to compare `sources` against `l0` + `compacted` based on the ordering but I think I may have complicated things a little bit by generating many collections to achieve it.  I agree the logic can be simplified and using the index to compare 1:1 may be a good idea. I still have some doubts regarding one case if we use the index to compare 1:1 strictly.  

Assuming `l0` = [SST-1,SST-2,SST-3] `compacted` = [100,50,0]

`sources` = [SST-2, SST-1, 100] -> 50  

Based on the [documentation](https://github.com/slatedb/slatedb/blob/main/docs/0002-compaction.md#compactions) this is a valid compaction but strictly comparing 1:1 will produce Invalid compaction because we would be missing SR 50 and 0 in the `source`. I can work around It if necessary but my main doubt here is understanding if this is a valid or invalid case as you said `sources` must have the same sequence of elements as the `l0` + `compacted`.

Reply:
For the index solution I have this prototype code, I haven't tested it yet but I just wanted to showcase the general idea. 
 
```rust
fn validate_compaction(l0: Vec<Ulid>, compacted:Vec<u32>, sources:Vec<SourceId>) -> Result<(), SlateDBError> {
        let (sst_idx,sr_idx) = match sources.first(){
            Some(SourceId::Sst(sst_id)) => {
                let mut index = l0.iter().position(|x| x == sst_id);
                (index,Some(0))
            },
            Some(SourceId::SortedRun(sr_id)) => {
                let mut index = compacted.iter().position(|x| x == sr_id);
                (None,index)
            },
            _ => (None,None)
        };

        if sst_idx.is_none() & sr_idx.is_none(){
            return Err(SlateDBError::InvalidCompaction);
        }
        let mut sst_idx = sst_idx.unwrap();
        let mut sr_idx = sr_idx.unwrap();

        let mut source_idx = 0usize;
        for i in sst_idx..l0.len(){
            let sst_id = match sources[source_idx]{
                SourceId::Sst(sst_id) => sst_id,
                _ => return Err(SlateDBError::InvalidCompaction)
            };

            if l0[i] != sst_id {
                return Err(SlateDBError::InvalidCompaction)
            }
            source_idx+=1;
        }

        for i in sr_idx..compacted.len(){
            let sr_id = match sources[source_idx]{
                SourceId::SortedRun(sr_id) => sr_id,
                _ => return Err(SlateDBError::InvalidCompaction) //todo: this should probably not raise an error but I will wait for clarification
            };

            if compacted[i] != sr_id {
                return Err(SlateDBError::InvalidCompaction)
            }
            source_idx+=1;
        }
        Ok(())
    }
```

Reply:
> Assuming l0 = [SST-1,SST-2,SST-3] compacted = [100,50,0]
> sources = [SST-2, SST-1, 100] -> 50

This would in fact not be a valid compaction, as it doesn't include 50 in the sources. If the compaction destination has the same id as one of the SRs in compacted, then the destination must be the last source in the sources list. Consider what would happen if we accepted this as a valid compaction. You're not including 50 in the sources being compacted, so it's data is removed from the database after compaction finishes.

I agree with @vigneshc that we can simplify:
- First, you can build up a list of `SourceId`s of everything in `l0` and `compacted` that represents the logical ordering of all the sources.
- Then, you need to verify that the sources from the compaction is a sub-list of this list.
- Finally, you can verify that the destination. I think we can simplify this part a bit for now. If the last compaction source is an l0 SST, then just verify that the destination is an SR with ID larger than the first element of `compacted`. If the last compaction source is an SR, then just verify the destination is the same as the last compaction source. This is a little stricter than the design doc, but that should be OK for now.

Reply:
I haven't been able to send the updates but will try to work through them today. Thanks for the comments.

Reply:
@brayanjuls Thanks for the update!